azkaban-db/src/main/java/azkaban/db/DatabaseOperator.java

```diff
@@ -117,16 +118,35 @@ public DatabaseOperator(final QueryRunner queryRunner) {
    * @return The number of rows updated.
    */
   public int update(final String updateClause, final Object... params) throws SQLException {
-    try {
-      return this.queryRunner.update(updateClause, params);
-    } catch (final SQLException ex) {
-      // todo kunkun-tang: Retry logics should be implemented here.
-      logger.error("update failed", ex);
-      if (this.dbMetrics != null) {
-        this.dbMetrics.markDBFailUpdate();
+    int retry = 0;
+    SQLException exception;
+    String errorMsg = "Update failed: Reached maximum number of retries: " + AzDBUtil.MAX_RETRIES_ON_DEADLOCK;
+    do {
+      try {
+        return this.queryRunner.update(updateClause, params);
+      } catch (final SQLException ex) {
+        exception = ex;
+        if (ex.getErrorCode() == 1213) { // 1213 -> Deadlock error
+          retry++;
+          logger.warn("Deadlock detected when trying to execute: " + updateClause + " with values: "
+              + Arrays.toString(params));
+          try {
+            Thread.sleep(AzDBUtil.RETRY_WAIT_TIME);
+          } catch (InterruptedException e) {
+            logger.info("Sleep interrupted");
+          }
+        } else {
+          errorMsg = "update failed";
```

Comment:
Can the error message be enhanced to provide more info about the error?

Reply:
This message is logged along with the SQL exception detailing the problem.